docs/intro/tutorial.rst

```diff
@@ -1,141 +1,446 @@
+    :class:`~web_poet.pages.WebPage` is a :ref:`generic class <generics>`,
+    which is why we could use ``WebPage[Book]`` in the definition of
+    ``BookPage`` to indicate ``Book`` as the output type of ``BookPage``.
+    However, ``BookPage`` is not a generic class, so something like
+    ``BookPage[CategorizedBook]`` would not work.
+
+    So instead you use :class:`~web_poet.pages.Returns`, a special, generic
+    class that you can inherit to re-define the output type of your page object
+    subclasses.
+
+After you update your ``tutorial-project/run.py`` script to request a
+``CategorizedBook`` item:
+
+.. literalinclude:: /tutorial-project/run.py
+   :language: python
+   :lines: 1-3, 5-8, 13-15, 17-18
+   :emphasize-lines: 4, 10
+
+And you execute it again:
+
+..  code-block:: bash
+
+    python tutorial-project/run.py
+
+You can see in the new output that your new classes have been used:
 
 .. code-block:: python
 
-    from web_poet.page_inputs import HttpResponse
+   CategorizedBook(title='The Exiled', category='Mystery', category_rank=None)
+
+
+Use additional requests
+=======================
+
+To extract data about an item, sometimes the HTTP response to a single URL is
+not enough. Sometimes, you need additional HTTP responses to get all the data
+that you want. That is the case with the ``category_rank`` attribute.
+
+The ``category_rank`` attribute indicates the position in which a book appears
+in the list of books of the category of that book. For example,
+`The Exiled`_ is 24th in the Mystery_ category, so the value of
+``category_rank`` should be ``24`` for that book.
+
+.. _The Exiled: http://books.toscrape.com/catalogue/the-exiled_247/index.html
+.. _Mystery: https://books.toscrape.com/catalogue/category/books/mystery_3/
 
+However, there is no indication of this value in the book details page. To get
+this value, you need to visit the URL of the category of the book whose
+data you are extracting, find the entry of that book within the grid of books
+of the category, and record in which position you found it. And categories with
+more than 20 books are split into multiple pages, so you may need more than 1
+additional request for some books.
 
-    response_data = HttpResponse(response.url,
-                                 body=response.content,
-                                 headers=response.headers)
-    page = BookLinksPage(response_data)
+Extend ``CategorizedBookPage`` in
+``tutorial-project/tutorial/pages/books_toscrape_com.py`` as follows:
 
-    print(page.to_item())
+.. literalinclude:: /tutorial-project/tutorial/pages/books_toscrape_com.py
+   :language: python
+   :lines: 14-15, 17, 19-26, 28-35, 39-52
+   :emphasize-lines: 1-3, 9, 11-12, 18-33
 
-Final Result
-============
+In the code above:
 
-Our simple Python script might look like this:
+-   You declare a new input in ``CategorizedBookPage``, ``http``, of type
+    :class:`~web_poet.page_inputs.client.HttpClient`.
```

Comment:
I wonder if this point could be more clear if we don't use WebPage, and use ItemPage directly. No strong opinon though.

Reply:
While I think you are right about its effect on this point, I think it is good for the tutorial to cover WebPage earlier, and showcase e.g. `self.css`. No strong opinion either, though.